src/structures/gdt.rs

```diff
@@ -306,6 +306,20 @@ impl Descriptor {
     /// Creates a TSS system descriptor for the given TSS.
     #[inline]
     pub fn tss_segment(tss: &'static TaskStateSegment) -> Descriptor {
+        // SAFETY: if iomap_size is zero, there are no requirements to uphold.
+        unsafe { Self::tss_segment_with_iomap(tss, 0) }
+    }
+
+    /// Creates a TSS system descriptor for the given TSS, setting up the IO permissions bitmap.
+    ///
+    /// # Safety
```

Comment:
I'm wondering if we could put enough checks here to make this method safe. We could take the IO map as a `[u8]` slice and then do all of the necessary checks (it's in range, byte past the end is all 1s, iomap_base is <= 0xDFFFH, etc..). We either need to do these checks or document all the requirements from the SDM in the `# Safety` section.

Reply:
Returning a result if it fails the checks?

Reply:
Either that or panic. We would need a custom error type for that (it could live in the tss module).

Reply:
Btw, I've expanded on the safety requirements. I couldn't find any others than terminating 0xff byte, iomap_base <= 0xdfff, and it points to the IO map slice, in the SDM.

Reply:
It might also be possible to have a `TssWithIoMap` struct as well, as discussed [here](https://github.com/rust-osdev/x86_64/pull/68#discussion_r279284959), but I'm unsure how non-8192-length io permissions bitmaps should be handled.

Reply:
Ya I think the idea would be:

In this PR, we have a function that takes two arguements a TSS and a bitmap slice, and does the necessary checks. This allows for custom sized iomaps of any size/position.

This would allow users to definite their own TssWithIoMap structure, and then just call the method we define here.

Reply:
Is the idea that the `tss_segment_with_iomap` function would set the `iomap_base` for the user?

Reply:
Later on, we could add a structure (the TssWithIoMap) to this library. That structure would make sure things stay valid (iomap_addr is correctly initialized, there's always a trailing 0xff byte, etc...) And we'd have a `descriptor` method on that structure that calls this method.

Reply:
I think it should just check that the iomap_base is correct, and return an error if it's wrong.

Reply:
Ok. I've pushed a prototype of this now.

Reply:
I'm thinking over this change again, and I think it has pretty good potential to make mutation of the IO map impossible. As I understand it, it should be possible to change the IO permissions bitmap at runtime, or am I mistaken? If so, giving &'static [u8] would make any modifications instant UB.

Reply:
The first part of this conversation seems to be resolved.



> I'm thinking over this change again, and I think it has pretty good potential to make mutation of the IO map impossible. As I understand it, it should be possible to change the IO permissions bitmap at runtime, or am I mistaken? If so, giving `&'static [u8]` would make any modifications instant UB. Perhaps it could take `&'static [UnsafeCell<u8>]` to get around this?

It's probably best to merge it as is and look into this in a separate PR. I'm not sure whether `UnsafeCell` would be enough for this or whether we need to e.g. use `AtomicU8`. We should ask that someone who has more knowledge of the Rust compiler.